### PR TITLE
Bug fix 3.4/shorter foot in door

### PR DESCRIPTION
**Pull request by mchacki**
This PR reduces the time an exclusive Lock is required during Synchronize-Finalize Phase.
This should speed up followers getting in sync in a cluster under high load.

Jenkins run:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1735/

### CODE REVIEW

File: arangod/Cluster/SynchronizeShard.cpp [hunk 1]
```diff
@@ -815,157 +819,248 @@ bool SynchronizeShard::first() {
...
-          cancelBarrier(ep, database, sy.get(BARRIER_ID).getNumber<int64_t>(), clientId);
-          if (longSync) {
-            LOG_TOPIC(ERR, Logger::MAINTENANCE)
-              << "synchronizeOneShard: long sync, after cancelBarrier"
-              << timepointToString(system_clock::now());
-          }
+      // From here on, we have to call `cancelBarrier` in case of errors
+      // as well as in the success case!
+      int64_t barrierId = sy.get(BARRIER_ID).getNumber<int64_t>();
+      TRI_DEFER(cancelBarrier(ep, database, barrierId, clientId));
```
Comment (graetzer) [inline arangod/Cluster/SynchronizeShard.cpp]: shouldn't you send cancelBarrier before logging you are done? I think you can use a scopeGuard and `guard.fire()` to trigger it manually

File: arangod/Cluster/SynchronizeShard.cpp [hunk 2]
```diff
@@ -815,157 +819,248 @@ bool SynchronizeShard::first() {
...
+  }
+
+  TRI_voc_tick_t tickReached = 0;
+  res = replicationSynchronizeCatchup(builder.slice(), tickReached);
+
+  if (!res.ok()) {
+    std::string errorMessage(
+      "synchronizeOneshard: error in syncCollectionCatchup: ") ;
+    errorMessage += res.errorMessage();
+    res = cancelReadLockOnLeader(ep, database, lockJobId, clientId, 60.0);
```
Comment (graetzer) [inline arangod/Cluster/SynchronizeShard.cpp]: shouldn't you do that via a scopeGuard as well for exception safety ?

File: arangod/Cluster/SynchronizeShard.cpp [hunk 3]
```diff
@@ -815,157 +819,248 @@ bool SynchronizeShard::first() {
...
+    if (!res.ok()) {
+      errorMessage += ", could not cancel read lock on leader: "
+        + res.errorMessage();
+    }
+    return {TRI_ERROR_INTERNAL, errorMessage};
+  }
+
+  // This result is unused, only in logs
+  res = cancelReadLockOnLeader(ep, database, lockJobId, clientId, 60.0);
```
Comment (graetzer) [inline arangod/Cluster/SynchronizeShard.cpp]: also do this via a scopeGuard

File: arangod/Cluster/SynchronizeShard.cpp (hunk 1 above)
Comment (mchacki) [reply arangod/Cluster/SynchronizeShard.cpp]: In this case it is actually fine with DEFER.
We are in a try catch scope, this is either left by error or exception => cancelBarrier
Or it is left successfully => cancelBarrier before logging for the user.

**Comment by mchacki** [comment]
Applied scopeGuards.

New Jenkins:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1756/

**Pull request merged**